SCons/Tool/MSCommon/common.py

```diff
@@ -256,19 +257,33 @@ def get_output(vcbat, args=None, env=None):
     env['ENV'] = normalize_env(env['ENV'], vs_vc_vars, force=False)
 
     if args:
-        debug("Calling '%s %s'" % (vcbat, args))
-        popen = SCons.Action._subproc(env,
-                                      '"%s" %s & set' % (vcbat, args),
-                                      stdin='devnull',
-                                      stdout=subprocess.PIPE,
-                                      stderr=subprocess.PIPE)
+        call_str = '"{}" {}'.format(vcbat, args)
     else:
-        debug("Calling '%s'" % vcbat)
+        call_str = '"{}"'.format(vcbat)
+
+    debug("Calling '%s'" % call_str)
+
+    call_str += ' & set'
+
+    # subprocess documentation notes:
+    #    The most common exception raised is OSError.
+    #    A ValueError will be raised if Popen is called with invalid arguments.
+    #    Exceptions defined subprocess module all inherit from SubprocessError.
+    try:
         popen = SCons.Action._subproc(env,
-                                      '"%s" & set' % vcbat,
+                                      call_str,
                                       stdin='devnull',
                                       stdout=subprocess.PIPE,
-                                      stderr=subprocess.PIPE)
+                                      stderr=subprocess.PIPE,
+                                      error='raise')
+    except OSError as e:
```

Comment:
what's the intent here?  catch a few specific exceptions and re-raise them, then catch essentially everything else and re-raise it too - and a bare `raise` would be plenty in any case.

Reply:
Simply an attempt at raising the same kind of exception with a reduced traceback to provide the most information possible.

As an example,
```
OSError: [WinError 2] The system cannot find the file specified:
```
versus
```
Exception: [WinError 2] The system cannot find the file specified:
```

Is there any utility in distinguishing between a ValueError (invalid subprocess arguments) and other exceptions?  

I can certainly remove all but the last clause if that is desired.  The three Exception types listed are the most likely exceptions raised by subprocess.

Reply:
Results from a "file-not-found" error in subprocess are shown below. 
Raising a MSVCScriptNotFound exception was manually commented out to allow subprocess to raise the error..

Both a bare raise:
```
try:
   call ...
except:
   raise
```
and a direct call without a try/except 
```
call ...
```
result in:
```
FileNotFoundError: [WinError 2] The system cannot find the file specified:
  File "S:\SCons\Test-Arg\SConstruct", line 29:
    env = Environment(MSVC_USE_SCRIPT=r'C:\test_batch-file_does-not_exist.bat')
  File "S:\SCons\scons\scripts\..\SCons\Environment.py", line 1010:
    apply_tools(self, tools, toolpath)
  File "S:\SCons\scons\scripts\..\SCons\Environment.py", line 116:
    _ = env.Tool(tool)
  File "S:\SCons\scons\scripts\..\SCons\Environment.py", line 1879:
    tool(self)
  File "S:\SCons\scons\scripts\..\SCons\Tool\__init__.py", line 273:
    self.generate(env, *args, **kw)
  File "S:\SCons\scons\scripts\..\SCons\Tool\default.py", line 41:
    SCons.Tool.Tool(t)(env)
  File "S:\SCons\scons\scripts\..\SCons\Tool\__init__.py", line 273:
    self.generate(env, *args, **kw)
  File "S:\SCons\scons\scripts\..\SCons\Tool\mslink.py", line 310:
    msvc_setup_env_once(env)
  File "S:\SCons\scons\scripts\..\SCons\Tool\MSCommon\vc.py", line 805:
    msvc_setup_env(env)
  File "S:\SCons\scons\scripts\..\SCons\Tool\MSCommon\vc.py", line 949:
    d = script_env(use_script)
  File "S:\SCons\scons\scripts\..\SCons\Tool\MSCommon\vc.py", line 747:
    stdout = common.get_output(script, args)
  File "S:\SCons\scons\scripts\..\SCons\Tool\MSCommon\common.py", line 274:
    error='raise')
  File "S:\SCons\scons\scripts\..\SCons\Action.py", line 797:
    pobj = subprocess.Popen(cmd, **kw)
  File "subprocess.py", line 729:
    
  File "subprocess.py", line 1017:
    
```
where:
```
try:
   call ...
except Exception as e:
   raise Exception(str(e))
```
is only about 6 lines shorter in this case:
```
Exception: [WinError 2] The system cannot find the file specified:
  File "S:\SCons\Test-Arg\SConstruct", line 29:
    env = Environment(MSVC_USE_SCRIPT=r'C:\test_batch-file_does-not_exist.bat')
  File "S:\SCons\scons\scripts\..\SCons\Environment.py", line 1010:
    apply_tools(self, tools, toolpath)
  File "S:\SCons\scons\scripts\..\SCons\Environment.py", line 116:
    _ = env.Tool(tool)
  File "S:\SCons\scons\scripts\..\SCons\Environment.py", line 1879:
    tool(self)
  File "S:\SCons\scons\scripts\..\SCons\Tool\__init__.py", line 273:
    self.generate(env, *args, **kw)
  File "S:\SCons\scons\scripts\..\SCons\Tool\default.py", line 41:
    SCons.Tool.Tool(t)(env)
  File "S:\SCons\scons\scripts\..\SCons\Tool\__init__.py", line 273:
    self.generate(env, *args, **kw)
  File "S:\SCons\scons\scripts\..\SCons\Tool\mslink.py", line 310:
    msvc_setup_env_once(env)
  File "S:\SCons\scons\scripts\..\SCons\Tool\MSCommon\vc.py", line 805:
    msvc_setup_env(env)
  File "S:\SCons\scons\scripts\..\SCons\Tool\MSCommon\vc.py", line 949:
    d = script_env(use_script)
  File "S:\SCons\scons\scripts\..\SCons\Tool\MSCommon\vc.py", line 747:
    stdout = common.get_output(script, args)
  File "S:\SCons\scons\scripts\..\SCons\Tool\MSCommon\common.py", line 276:
    raise Exception(str(e))
```
Other than the length of the traceback, is there a compelling reason to even have the try/except block?

I'm happy to make whatever changes that are deemed best in this case.

Reply:
Have to think about it.  In this case we could instead raise an SCons error, which will reduce the call stack. Although maybe a little too terse?
```python
    except FileNotFoundError as e:
        raise SCons.Errors.UserError(e) from None
```

Reply:
The FileNotFoundError *should* be prevented by the code added to check that MSVC_USE_SCRIPT exists.  The not found above is just to test what happens when subprocess returns an unexpected exception triggered by commenting out the code that checks MSVC_USE_SCRIPT.  

It may not be a user error per se but possibly a (transient) system problem locally with the subprocess stack (i.e., valid arguments to the call but the call fails anyway).

Reply:
Yes, it's a UserError only if we can prove the SConscript requested a non-existent file, not if something went wrong on the system.  It was just an illustration I could get to trigger to make sure I didn't have the syntax wrong :)

Reply:
A bit off topic: an enhancement that I was planning on proposing once this was resolved is the addition of a user specified script argument: MSVC_USE_SCRIPT_ARG to be used in conjunction with MSVC_USE_SCRIPT.  This would allow any flags recognized by the msvc batch files to be passed along with the script name.

For example, this would allow the following which currently can't be achieved directly in a SConstruct/SConscript file without resorting to a user-defined batch file (VS2022 Preview, amd64 target, toolset 14.2, sdk 10.0.19041.0):
```
use_version    = '14.3'
use_script     = 'c:\\software\\msvs143bt-pre\\VC\\Auxiliary\\Build\\vcvarsall.bat'
use_script_arg = 'amd64 --vcvars_ver=14.2 10.0.19041.0'

env = Environment(MSVC_VERSION = use_version, MSVC_USE_SCRIPT = use_script, MSVC_USE_SCRIPT_ARG = use_script_arg)
```

If something like this were added, there would be no good way to validate the user provided script argument other than call subprocess.  In this case it could be harder to distinguish between a user error and a system error.  

Unfortunately, current msvc logic will not pick up batch file errors correctly (unless cl.exe is not added to the path).  There is room for improvement in this area.

Reply:
```python
>>> "room for improvement" in msvc_tool
True
```
:-)

Reply:
Further off-topic: If a user script argument were allowed and an overloaded version when MSVC_USE_SCRIPT=None allowing a dictionary to be passed, it is possible in user space to unify/resolve most of the outstanding msvc issues/requests (preview, toolsets, components, etc).

Working SConstruct fragments via script argument enhancements:
```
from vsdetect import scons_msvc_byinstance as scons_msvc

# SCons runs batch file: MSVC_USE_SCRIPT (string) and MSVC_USE_SCRIPT_ARG (string)
env = Environment(**scons_msvc(MSVC_VERSION='14.3', MSVC_PREVIEW=True, MSVC_COMPONENT='BuildTools', MSVC_TOOLSET='14.2', TARGET_ARCH='amd64', MSVC_ADD_ARG='10.0.19041.0', run_script=False))

# user runs batch file: MSVC_USE_SCRIPT (None) and MSVC_USE_SCRIPT_ARG (dictionary)
env = Environment(**scons_msvc(MSVC_VERSION='14.3', MSVC_PREVIEW=True, MSVC_COMPONENT='BuildTools', MSVC_TOOLSET='14.2', TARGET_ARCH='amd64', MSVC_ADD_ARG='10.0.19041.0', run_script=True))

# lookup by c runtime
env = Environment(**scons_msvc(MSVC_RUNTIME='140', MSVC_PREVIEW=True, MSVC_COMPONENT='BuildTools', MSVC_TOOLSET='14.2', TARGET_ARCH='amd64', MSVC_ADD_ARG='10.0.19041.0', run_script=True))

# lookup by toolset
env = Environment(**scons_msvc(MSVC_TOOLSET='14.2', MSVC_PREVIEW=True, MSVC_COMPONENT='BuildTools', TARGET_ARCH='amd64', MSVC_ADD_ARG='10.0.19041.0', run_script=True))
```

This would also allow user-side caching if the overloaded dictionary support were added.

Reply:
@jcbrill - can you file an issue about additional arguments to vcvarsall.bat, I have some additional thoughts which would be good to capture in such an issue rather than here.